typegen/evaluation/file_type_hints_collector.py

```diff
@@ -0,0 +1,159 @@
+import os.path
+import pathlib
+from typing import Dict
+import pandas as pd
+import libcst as cst
+from libcst.metadata import PositionProvider
+import constants
+from tracing import TraceDataCategory
+
+
+class FileTypeHintsCollector:
+    """Collects the type hints of multiple .py files."""
+    def __init__(self, project_dir: pathlib.Path):
+        self.project_dir = project_dir
+        self.typehint_data = pd.DataFrame()
+        self.file_path = ""
+
+    def collect_data(self, file_paths: list[pathlib.Path]) -> None:
+        """Collects the type hints of the provided file paths."""
+        for file_path in file_paths:
+            with file_path.open() as file:
+                file_content = file.read()
+
+            module = cst.parse_module(source=file_content)
+            module_and_meta = cst.MetadataWrapper(module)
+            relative_path = str(pathlib.Path(os.path.relpath(file_path.resolve(), self.project_dir.resolve())))
+            visitor = _TypeHintVisitor(relative_path)
+            module_and_meta.visit(visitor)
+            typehint_data = visitor.typehint_data.astype(constants.TraceData.TYPE_HINT_SCHEMA)
+
+            self.typehint_data = pd.concat(
+                [self.typehint_data, typehint_data], ignore_index=True
+            ).astype(constants.TraceData.TYPE_HINT_SCHEMA)
+
+
+class _TypeHintVisitor(cst.CSTVisitor):
+    METADATA_DEPENDENCIES = (PositionProvider,)
+
+    def __init__(self, file_path: str) -> None:
+        super().__init__()
+        self.file_path = file_path
+        self.typehint_data = pd.DataFrame(columns=constants.TraceData.TYPE_HINT_SCHEMA.keys())
+        self._scope_stack: list[cst.FunctionDef | cst.ClassDef] = []
+        self.imports: Dict[str, str] = {}
+
+    def _innermost_class(self) -> cst.ClassDef | None:
+        fromtop = reversed(self._scope_stack)
+        classes = filter(lambda p: isinstance(p, cst.ClassDef), fromtop)
+
+        first: cst.ClassDef | None = next(classes, None)  # type: ignore
+        return first
+
+    def _innermost_function(self) -> cst.FunctionDef | None:
+        fromtop = reversed(self._scope_stack)
+        fdefs = filter(lambda p: isinstance(p, cst.FunctionDef), fromtop)
+
+        first: cst.FunctionDef | None = next(fdefs, None)  # type: ignore
+        return first
+
+    def visit_ImportFrom(self, node: cst.ImportFrom) -> bool | None:
+        module_name: str = str(node.module.value)  # type: ignore
+        for name in node.names:  # type: ignore
```

Comment:
See below: the case `from x import y` and `y` is not a class is not covered here.
But try it in a test, if I am wrong, then this is OK

Reply:
Bonus: all imports relevant for type hinting are in a `if TYPE_CHECKING:` block close to the top of the file; only consider these imports instead of all imports the files has.

Reply:
As per yesterday's discord discussion:

Our generated file will only contain type hints that have imports from the  `if TYPE_CHECKING:` block. 
Therefore, all type hints in the generated file must be connected to their importing statement from within this block. What I am worried about is that an already existing import in the file contains a symbol by the same name that would make the imported class ambiguous.
The fix for this is very simple, and could preemptively save us a lot of bug-tracking. 
The potential reward to effort ratio is too high to not consider doing this.

Reply:
There is no such constraint that files contain this if TYPE_CHECKING block. Additionally, all imports are important. They provide the information of the module name of a class name. It is not possible to skip any import.

However, the issue of ambiguity might be a valid bug. I will test it.